### PR TITLE
Web trackpad pan

### PR DESCRIPTION
**Pull request by moffatman**
Use a kind of heuristic to guess at whether an event is from mouse or trackpad.

Part of https://github.com/flutter/flutter/issues/64210

## Sequence

1. https://github.com/flutter/flutter/pull/112171
2. This PR

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 1]
```diff
@@ -324,6 +324,12 @@ mixin _WheelEventListenerMixin on _BaseAdapter {
     const int domDeltaLine = 0x01;
     const int domDeltaPage = 0x02;
 
+    // Logic from https://stackoverflow.com/a/62415754
+    ui.PointerDeviceKind kind = ui.PointerDeviceKind.mouse;
+    if (event.deltaX % 120 != 0 || event.deltaY % 120 != 0) {
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/pointer_binding.dart:329]: Where did you get the 120 value?  Maybe add a comment explaining that (like what device and browser, if that's what you did).

**Review by justinmc — COMMENTED** [review]
Sorry I did miss this PR among the other trackpad PRs!

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 2]
```diff
@@ -324,6 +324,12 @@ mixin _WheelEventListenerMixin on _BaseAdapter {
     const int domDeltaLine = 0x01;
     const int domDeltaPage = 0x02;
 
+    // Logic from https://stackoverflow.com/a/62415754
```
[the remark this answers is not in the document]
Comment (moffatman) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:327]: It isn't copied, just where there is a lot of relevant discussion, so I included the link for reference. But I can remove it.

File: lib/web_ui/lib/src/engine/pointer_binding.dart (hunk 1 above)
Comment (moffatman) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:329]: It's just a convention which dates from Windows I think, not really documented but it holds on all major browsers. I'll update the comment.

Comment (justinmc) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:328]: Neat, I never knew that 👍

**Review by justinmc — APPROVED** [review]
LGTM 👍 

CC @mdebbar in case you or anyone else more familiar with the web engine wants to take a look before this is merged.

Edit: Just noting that this PR shouldn't be merged until the framework PR is merged.  I tried them out together and left a [review](https://github.com/flutter/flutter/pull/112171#pullrequestreview-1157412333) over there with some problems I ran into.

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 3]
```diff
@@ -340,6 +340,24 @@ mixin _WheelEventListenerMixin on _BaseAdapter {
...
+      // is not divisible by 120, this event is not a normal mouse wheel event.
+      // On macOS, mouse wheel events by default have an acceleration curve applied,
+      // so their delta values ramp up and are not at fixed multiples of 120.
+      // But in this case, the wheelDelta properties of the event still keep
+      // their original values.
+      // For all events without this acceleration curve applied, the wheelDelta
+      // values are by convention three times greater than the delta values and with
+      // the opposite sign.
+      kind = ui.PointerDeviceKind.trackpad;
+    }
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/pointer_binding.dart:359]: Suggestion: extract this logic into a function called `isTrackpadEvent` or similar, to make it more readable.

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 4]
```diff
@@ -340,6 +340,24 @@ mixin _WheelEventListenerMixin on _BaseAdapter {
...
+      // While not standardized, `blink` and `webkit` browsers use a delta of 120 to
+      // represent one mouse wheel turn. If either dimension of the delta
+      // is not divisible by 120, this event is not a normal mouse wheel event.
+      // On macOS, mouse wheel events by default have an acceleration curve applied,
+      // so their delta values ramp up and are not at fixed multiples of 120.
+      // But in this case, the wheelDelta properties of the event still keep
+      // their original values.
+      // For all events without this acceleration curve applied, the wheelDelta
+      // values are by convention three times greater than the delta values and with
+      // the opposite sign.
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/pointer_binding.dart:357]: It would be nice to add links to back up some of these statements.

**Review by mdebbar — APPROVED** [review]
PR looks good to me with some minor suggestions.

**Comment by Piinks** [comment]
I'd like to test this out on Monday, I have been working on animating discrete mouse wheel data in https://github.com/flutter/flutter/pull/115314, and ran into the fact that we don't distinguish between trackpad and mouse on web. 
I've been catching up on the discussion in https://github.com/flutter/flutter/pull/112171#issuecomment-1315692789 as well.
I have a **_bunch_** of different mice and trackpads I have been testing with, so I can do the same here. Thanks for contributing this! This may enable a lot of different fixes in the framework!

**Comment by skia-gold** [comment]
Gold has detected about 70 new digest(s) on patchset 9.
View them at https://flutter-engine-gold.skia.org/cl/github/36346

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 9.
View them at https://flutter-engine-gold.skia.org/cl/github/36346

**Comment by moffatman** [comment]
Ah I think the PR's Gold is broken now, I had the issue before, will probably have to recreate it...

**Comment by Piinks** [comment]
> Ah I think the PR's Gold is broken now, I had the issue before, will probably have to recreate it...

I think you can resolve it by creating a new commit and not force pushing. Skia Gold is easily confused by force pushes unfortunately.

**Comment by Piinks** [comment]
Oh it looks like Gold is fine. 👍

**Comment by Piinks** [comment]
Ok I have tested this out with 4 mice and 2 trackpads running web on Mac. 😁 

IMG_0434

I ran a simple flutter app with an infinite number of scrolling items in a list, and had (in the framework) the event data output to tell me:

1. trackpad or mouse
2. time stamp
3. unconverted scroll delta (before it is divided by the device pixel ratio)
4. final scroll delta reported to the framework

For trackpads, every single one was correctly identified as a trackpad. 🎉 

For the 4 mice, all of them were able to fool this into thinking it was a trackpad when I scrolled the wheel really fast. It was easier to do on some compared to others, and it would only occur once per occasion - the event before and after were correctly identified as mouse. The incoming delta was ~40 or ~320 in the instances I saw it occur.

On one of the cheap mouses I have, it was really easy to hit the 40 delta, so it was easier to reproduce. On one of the gaming mice, it was harder to trigger it, and would happen with an input of 320. I am guessing these devices all have a different level of precision.

Here is some of that output, let me know if it helps, or if I can gather more info for you. :)
This is one of the cheap mice that was really easy to trigger with a fling on the wheel.

```
Unconverted delta: 4.000244140625
PointerDeviceKind.mouse - 0:15:28.540299 : Offset(-0.0, 4.0)
PointerDeviceKind.mouse
Unconverted delta: 40.0189208984375
PointerDeviceKind.trackpad - 0:15:28.556200 : Offset(-0.0, 40.0)
PointerDeviceKind.trackpad
Unconverted delta: 168.1787109375
PointerDeviceKind.mouse - 0:15:28.564299 : Offset(-0.0, 168.2)
PointerDeviceKind.mouse
Unconverted delta: 500.9765625
PointerDeviceKind.mouse - 0:15:28.580200 : Offset(-0.0, 501.0)
PointerDeviceKind.mouse
Unconverted delta: 684.9090576171875
PointerDeviceKind.mouse - 0:15:28.596200 : Offset(-0.0, 684.9)
PointerDeviceKind.mouse
Unconverted delta: 365.3118896484375
PointerDeviceKind.mouse - 0:15:28.604200 : Offset(-0.0, 365.3)
PointerDeviceKind.mouse
```

**Comment by moffatman** [comment]
I was able to reproduce it, looks like sometimes macOS scroll acceleration can by chance happen to line up with expected values for a trackpad. Should be solvable with by cross-checking with the previous event, since this shouldn't happen for more than one event in a row.

**Comment by moffatman** [comment]
@Piinks See if you have the same problems with the latest change

**Comment by Piinks** [comment]
Thanks! I will check it out today!

**Comment by Piinks** [comment]
Still able to trip it at ~40:

```
Engine says: PointerDeviceKind.mouse
Unconverted delta: 4.000244140625
PointerDeviceKind.mouse - 0:00:05.808799 : Offset(-0.0, 4.0)
PointerDeviceKind.mouse
Engine says: PointerDeviceKind.trackpad
Unconverted delta: 39.991455078125
PointerDeviceKind.trackpad - 0:00:05.824900 : Offset(-0.0, 40.0)
PointerDeviceKind.trackpad
Engine says: PointerDeviceKind.mouse
Unconverted delta: 159.752197265625
PointerDeviceKind.mouse - 0:00:05.840900 : Offset(-0.0, 159.8)
PointerDeviceKind.mouse
Engine says: PointerDeviceKind.mouse
Unconverted delta: 229.3438720703125
PointerDeviceKind.mouse - 0:00:05.848900 : Offset(-0.0, 229.3)
PointerDeviceKind.mouse
Engine says: PointerDeviceKind.mouse
Unconverted delta: 1211.998291015625
PointerDeviceKind.mouse - 0:00:05.888900 : Offset(-0.0, 1212.0)
PointerDeviceKind.mouse
```

Let me know if there is more info I can gather.

**Comment by moffatman** [comment]
Can you capture the issue again with
```dart
print('deltaX=${event.deltaX}, deltaY=${event.deltaY}, wheelDeltaX=${event.wheelDeltaX}, wheelDeltaY=${event.wheelDeltaY}, timeStamp=${event.timeStamp}');
```
at the beginning of `_isTrackpadEvent` in `pointer_binding.dart`? Also would be good to see at least 2 events before the incorrect one.

**Comment by Piinks** [comment]
> Can you capture the issue again with ...

Totally, whipping this up now. I was out for the holiday.

> Also would be good to see at least 2 events before the incorrect one.

In the examples I shared, that is the whole event, from starting to ending the scroll on the stepped wheel. The incorrect one showed up on the second event.

**Comment by Piinks** [comment]
Here it is with one quick twirl on the mouse wheel:

```
Performing hot restart...                                          153ms
Restarted application in 153ms.
deltaX=-0.0, deltaY=4.000244140625, wheelDeltaX=0, wheelDeltaY=-120, timeStamp=27866.300000190735
Mouse
deltaX=-0.0, deltaY=39.9945068359375, wheelDeltaX=0, wheelDeltaY=-120, timeStamp=27882.200000286102
Trackpad
deltaX=-0.0, deltaY=433.90625, wheelDeltaX=0, wheelDeltaY=-1200, timeStamp=27906.200000286102
Mouse
deltaX=-0.0, deltaY=654.576416015625, wheelDeltaX=0, wheelDeltaY=-1800, timeStamp=27922.300000190735
Mouse
deltaX=-0.0, deltaY=764.7760009765625, wheelDeltaX=0, wheelDeltaY=-2160, timeStamp=27938.200000286102
Mouse
```

**Comment by Piinks** [comment]
Nice! I just tested it with all of the devices I have, it was correct every time. 🎉

**Review by Piinks — APPROVED** [review]
@yjbanov was also interested in this.. FYI

This LGTM!

**Comment by yjbanov** [comment]
Thanks @Piinks for jumping in and testing it thoroughly! Great to see progress made. I'm curious what failures would look and feel like if this logic fails. We can't enumerate all browser engines, and some browser engines may be customized to use different numbers. In case case, will we observe slight degradations in scrolling quality? Or could the issues be severe?

**Comment by moffatman** [comment]
This doesn't affect scrolling, it's only for custom interactions such as InteractiveViewer. If there was some problem you might see strangeness there where during a pan it might zoom by accident on one event in a series. We can gate it to only certain browser / OS combinations if you want. Already Firefox is ruled out. I personally have only verified it on macOS Chrome and Safari.

**Comment by Piinks** [comment]
Thanks for explaining @moffatman. I plan to use this in the framework to execute an animation when mouse wheel scroll input comes from a mouse, ignoring trackpad - as an additional example of how this will be used. From my understanding, this change will not necessarily change anything in the framework unless we make a change there based on the input device.

**Comment by yjbanov** [comment]
Cool, since the scrolling changes will roll out as opt-in, the risk is not as big, so let's just keep an eye on it. LGTM!

**Review by yjbanov — APPROVED** [review]

**Comment by yjbanov** [comment]
@moffatman would you mind rebasing the PR on top of `main` to kick the Mac Host Engine bot?

**Comment by moffatman** [comment]
@Piinks I updated the logic again, can you recheck with the various devices you have and see if there is a regression? Thanks 😄

**Comment by Piinks** [comment]
Sure thing! I have even more mice now. 😳

**Comment by Piinks** [comment]
Tested with 7 mice and 2 trackpads. I was not able to fool it. @justinmc had different luck last time with the magic mouse. On my magic mouse, I swiped and flung on it a bunch but did not see it come through as a mouse. :+1:

**Comment by justinmc** [comment]
@Piinks That's great!  I'll try again with my magic mouse tonight and then plan to merge https://github.com/flutter/flutter/pull/112171 if I can't reproduce that bug.

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 5]
```diff
@@ -264,6 +264,7 @@ abstract class _BaseAdapter {
   final PointerDataConverter _pointerDataConverter;
   final KeyboardConverter _keyboardConverter;
   DomWheelEvent? _lastWheelEvent;
+  bool? _lastWheelEventWasTrackpad;
```
[the remark this answers is not in the document]
Comment (moffatman) [reply lib/web_ui/lib/src/engine/pointer_binding.dart:267]: I have it this way as in some previous iterations the null state did matter.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

I left a review on the framework PR (https://github.com/flutter/flutter/pull/112171) saying that I was unable to break this using the Magic Mouse this time.  Thanks for fixing that and for cleaning up those last few comments!

Rerunning the Fuchsia test, it looked like an infrastructure flake.

**Comment by Piinks** [comment]
The framework change landed, I am just waiting on a lean build from that in the framework and then we can land this. 🎉

File: lib/web_ui/lib/src/engine/pointer_binding.dart [hunk 6]
```diff
@@ -333,13 +335,83 @@ abstract class _BaseAdapter {
...
+      return false;
+    }
+    if (_isAcceleratedMouseWheelDelta(event.deltaX, event.wheelDeltaX) ||
+        _isAcceleratedMouseWheelDelta(event.deltaY, event.wheelDeltaY)) {
+      return false;
+    }
+    if (((event.deltaX % 120 == 0) && (event.deltaY % 120 == 0)) ||
+        (((event.wheelDeltaX ?? 1) % 120 == 0) && ((event.wheelDeltaY ?? 1) % 120) == 0)) {
+      // While not in any formal web standard, `blink` and `webkit` browsers use
+      // a delta of 120 to represent one mouse wheel turn. If both dimensions of
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/pointer_binding.dart:375]: https://github.com/chromium/chromium/blob/fd8a8914ca0183f0add65ae55f04e287543c7d4a/ui/events/event.cc#L676-L684

**Pull request merged**